Live-Snap/Live-Snap/Controllers/FetchSnapcodeViewController.swift

```diff
@@ -182,22 +253,77 @@ class FetchSnapcodeViewController: UIViewController, UITextFieldDelegate {
                 
                 if let bitmojiSnapcodeImage = bitmojiSnapcodeImage {
-                    System.shared.snapcode = bitmojiSnapcodeImage
+                    
+                    if let snapcodeImage = snapcodeImage {
+                        System.shared.snapcode = bitmojiSnapcodeImage
+                        self.bitmojiSnapcode = bitmojiSnapcodeImage
+                        self.defaultSnapcode = snapcodeImage
+                        self.snapcodeSelectionUI()
+                    } else {
+                        self.fetchSnapcodeIndicator.dismiss(animated: false)
+                        self.fetchSnapcodeIndicator.animation.animationFinished()
+                        System.shared.snapcode = bitmojiSnapcodeImage
+                        self.nextButtonWasPressed()
+                    }
+                    
                 } else if let snapcodeImage = snapcodeImage {
                     System.shared.snapcode = snapcodeImage
+                    self.fetchSnapcodeIndicator.dismiss(animated: false)
+                    self.fetchSnapcodeIndicator.animation.animationFinished()
+                    self.nextButtonWasPressed()
                 } else {
                     self.showCannotGetSnapcodeError()
-                    self.nextButton.isUserInteractionEnabled = true
                     return
                 }
                 
-                DispatchQueue.main.async {
-                    
-                    let selectPhotoViewController = SelectPhotoViewController()
-                    System.shared.appDelegate().pageViewController?.setViewControllers([selectPhotoViewController], direction: .forward, animated: true, completion: nil)
-                }
             }
         }
     }
     
+    func snapcodeSelectionUI() {
+        fetchSnapcodeIndicator.dismiss(animated: false)
+        fetchSnapcodeIndicator.animation.animationFinished()
+        
+        snapcodeImageView.image = bitmojiSnapcode
+        
+        usernameTextField.isUserInteractionEnabled = false
+        usernameTextField.frame.origin.y = snapcodeImageView.frame.maxY + 40
+        
+        let hapticFeedback = UIImpactFeedbackGenerator(style: .heavy)
+        hapticFeedback.impactOccurred()
+        
+        view.addSubview(snapcodeSelectionToolbar)
+        view.addSubview(toolbarInsturctionsLabel)
+        view.addSubview(usernameTextField)
+        view.addSubview(snapcodeSelectionInsturctionsLabel)
+        view.addSubview(snapcodeSelectionSwitch)
+        view.addSubview(snapcodeImageView)
+    }
+    
+    @objc func switchIsTapped(toggle: UISwitch) {
+        if toggle.isOn {
+            snapcodeImageView.image = bitmojiSnapcode
+            System.shared.snapcode = bitmojiSnapcode
+        } else {
+            snapcodeImageView.image = defaultSnapcode
+            System.shared.snapcode = defaultSnapcode
+        }
+    }
+    
+    @objc func cancelButtonWasPressed() {
+        DispatchQueue.main.async {
```

Comment:
unless this is being called from a non UI thread, you don't need to dispatch to the main thread